### ISSUE TITLE
cuda runtime error (59) : device-side assert triggered

### ISSUE THREAD

**Issue by richardassar** [torch/cutorch#708]
With code that was previously working fine until a recent Torch reinstall I'm now receiving the following error...

> /home/richard/torch/extra/cutorch/lib/THC/THCTensorIndex.cu:321: void indexSelectLargeIndex(TensorInfo<T, IndexType>, TensorInfo<T, IndexType>, TensorInfo<long, IndexType>, int, int, IndexType, IndexType, long) [with T = float, IndexType = unsigned int, DstDim = 2, SrcDim = 2, IdxDim = -2]: block: [49,0,0], thread: [0,0,0] Assertion `srcIndex < srcSelectDimSize` failed.

...

> THCudaCheck FAIL file=/home/richard/torch/extra/cutorch/lib/THC/generic/THCTensorCopy.c line=18 error=59 : device-side assert triggered
> /home/richard/torch/install/bin/luajit: .../richard/torch/install/share/lua/5.1/threads/threads.lua:183: [thread 2 callback] /home/richard/torch/install/share/lua/5.1/nn/Container.lua:67: 
> In 3 module of nn.Sequential:
> In 1 module of nn.ConcatTable:
> In 2 module of nn.Sequential:
> In 1 module of nn.Sequential:
> In 2 module of nn.ParallelTable:
> In 1 module of nn.Sequential:
> /home/richard/torch/install/share/lua/5.1/nn/Bottle.lua:33: cuda runtime error (59) : device-side assert triggered at /home/richard/torch/extra/cutorch/lib/THC/generic/THCTensorCopy.c:18
> stack traceback:
>         [C]: in function 'copy'
>         /home/richard/torch/install/share/lua/5.1/nn/Bottle.lua:33: in function </home/richard/torch/install/share/lua/5.1/nn/Bottle.lua:16>

Assert introduced here:

https://github.com/torch/cutorch/commit/c58f2f2f6e9f3c8443058d226fc37b29c7a16fb3#diff-b27a6c2ae9fc16f5dfb01cf6cb45bb82R321

The copy:

https://github.com/torch/cutorch/blob/master/lib/THC/generic/THCTensorCopy.c#L18

Invoked from:

https://github.com/torch/nn/blob/master/Bottle.lua#L33

Which is just a small shape tensor being copied.

My code feeds fixed size [batch x seq x feature] dim tensor each iteration and upon the 2486th I receive the above error. I reproduced this multiple times and it always happened exactly on this iteration, strangely.

If I back out c58f2f2f6e9f3c8443058d226fc37b29c7a16fb3 the problem "goes away" but obviously this isn't ideal.

Anyone else seeing this?

**Comment by richardassar** [comment]
I also get the following

```
/home/richard/torch/extra/cunn/lib/THCUNN/ClassNLLCriterion.cu:52: void cunn_ClassNLLCriterion_updateOu
tput_kernel(Dtype *, Dtype *, Dtype *, long *, Dtype *, int, int, int, int) [with Dtype = float, Acctyp
e = float]: block: [0,0,0], thread: [31,0,0] Assertion `t >= 0 && t < n_classes` failed.
THCudaCheck FAIL file=/home/richard/cutorch/lib/THC/generic/THCStorage.c line=32 error=59 : device-side
 assert triggered
/home/richard/torch/install/bin/luajit: cuda runtime error (59) : device-side assert triggered at /home
/richard/cutorch/lib/THC/generic/THCStorage.c:32
stack traceback:
        [C]: at 0x7fa84662e670
        [C]: in function '__index'
        ...ard/torch/install/share/lua/5.1/nn/ClassNLLCriterion.lua:52: in function 'updateOutput'
        ...ard/torch/install/share/lua/5.1/nn/ParallelCriterion.lua:23: in function 'forward'
        train.lua:315: in function 'opfunc'
        /home/richard/torch/install/share/lua/5.1/optim/adam.lua:37: in function 'adam'
        train.lua:338: in function 'train'
        train.lua:443: in main chunk
        [C]: in function 'dofile'
        ...hard/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
        [C]: at 0x00405e90
```

However I have found a 0 that is slipping into my targets which might at least explain this one as t would be -1

**Comment by soumith** [comment]
we've enabled device side asserts recently, that's why you see existing bugs in your code now exposed (they were silently ignored before). Because of the asynchronous nature of cuda, the assert might not point to a full correct stack trace pointing to where the assert was triggered from.
if you run the program with CUDA_LAUNCH_BLOCKING=1 th script.lua
this will help get a more exact stack trace

**Comment by richardassar** [comment]
Thanks for your help Soumith. 

It looks like my threaded audio loader / mini-batch maker is the culprit for the 0 (class index) but the nn.Bottle issue remains elusive. I'll comment if I resolve the issue on my side with latest cutorch, in case it helps anyone else who hits a similar issue.

**Comment by soumith** [comment]
for the botte issue, see the last line of my comment above, will help debug

**Comment by richardassar** [comment]
Was just about to say, if the true assert triggered was due to the bad 0-class entry in my targets then I assume the nn.Bottle assert was a false positive. 

Running with `CUDA_LAUNCH_BLOCKING=1` now, thanks again.

**Comment by richardassar** [comment]
Just confirming that this is resolved, the nn.Bottle assert was not the true failing assert. It would be good to optionally catch these host-side before they manifest on the device as in some cases it could be extremely time consuming to track down the true source with kernel launch blocking.

In this case a zero that sneaked into my targets was invalid input for ClassNLLCriterion.

**Comment by soumith** [comment]
due to cuda 's asynchronous nature, we cannot catch these host side without a significant drop in performance

**Comment by richardassar** [comment]
True, it is costly to validate host side.

Perhaps you could wrap assert to printf (or write to a custom ring buffer)
that is dumped before exiting Torch? You'd have to synchronise to gather
all output.

This would accumulate the output from every kernel with an assert failure
rather than just displaying one at random.

In any case I solved my issue. Thanks again.

**Comment by pavanky** [comment]
one potential alternative would be to have a reduction + assert before calling the kernel. Doing this all the time is going to be costly (and may not even be feasible all the time), but it could be useful if it could be optionally enabled to help with debugging.

**Comment by Larriers** [comment]
I also get the similar following error
```
/tmp/luarocks_cunn-scm-1-9248/cunn/lib/THCUNN/ClassNLLCriterion.cu:52: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long *, Dtype *, int, int, int, int) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [13,0,0] Assertion `t >= 0 && t < n_classes` failed.
THCudaCheck FAIL file=/tmp/luarocks_cunn-scm-1-9248/cunn/lib/THCUNN/generic/ClassNLLCriterion.cu line=83 error=59 : device-side assert triggered
luajit: /usr/local/share/lua/5.1/nn/THNN.lua:110: cuda runtime error (59) : device-side assert triggered at /tmp/luarocks_cunn-scm-1-9248/cunn/lib/THCUNN/generic/ClassNLLCriterion.cu:83
stack traceback:
	[C]: in function 'v'
	/usr/local/share/lua/5.1/nn/THNN.lua:110: in function 'ClassNLLCriterion_updateOutput'
	/usr/local/share/lua/5.1/nn/ClassNLLCriterion.lua:44: in function 'forward'
	/home/amax/ceshi0415/daima/opensource_base.lua:528: in function 'train_epoch'
	opensource_baseline.lua:54: in function 'runTrainVal'
	opensource_baseline.lua:143: in main chunk
	[C]: at 0x004061f0

```
@soumith I have tried run the program with CUDA_LAUNCH_BLOCKING=1, however, it also presents the following error:
```
/tmp/luarocks_cunn-scm-1-9248/cunn/lib/THCUNN/ClassNLLCriterion.cu:52: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long *, Dtype *, int, int, int, int) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [3,0,0] Assertion `t >= 0 && t < n_classes` failed.
/tmp/luarocks_cunn-scm-1-9248/cunn/lib/THCUNN/ClassNLLCriterion.cu:52: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long *, Dtype *, int, int, int, int) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [7,0,0] Assertion `t >= 0 && t < n_classes` failed.
/tmp/luarocks_cunn-scm-1-9248/cunn/lib/THCUNN/ClassNLLCriterion.cu:52: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long *, Dtype *, int, int, int, int) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [9,0,0] Assertion `t >= 0 && t < n_classes` failed.
/tmp/luarocks_cunn-scm-1-9248/cunn/lib/THCUNN/ClassNLLCriterion.cu:52: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long *, Dtype *, int, int, int, int) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [13,0,0] Assertion `t >= 0 && t < n_classes` failed.
/tmp/luarocks_cunn-scm-1-9248/cunn/lib/THCUNN/ClassNLLCriterion.cu:52: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long *, Dtype *, int, int, int, int) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [22,0,0] Assertion `t >= 0 && t < n_classes` failed.
/tmp/luarocks_cunn-scm-1-9248/cunn/lib/THCUNN/ClassNLLCriterion.cu:52: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long *, Dtype *, int, int, int, int) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [30,0,0] Assertion `t >= 0 && t < n_classes` failed.
THCudaCheck FAIL file=/tmp/luarocks_cunn-scm-1-9248/cunn/lib/THCUNN/generic/ClassNLLCriterion.cu line=83 error=59 : device-side assert triggered
luajit: /usr/local/share/lua/5.1/nn/THNN.lua:110: cuda runtime error (59) : device-side assert triggered at /tmp/luarocks_cunn-scm-1-9248/cunn/lib/THCUNN/generic/ClassNLLCriterion.cu:83
stack traceback:
	[C]: in function 'v'
	/usr/local/share/lua/5.1/nn/THNN.lua:110: in function 'ClassNLLCriterion_updateOutput'
	/usr/local/share/lua/5.1/nn/ClassNLLCriterion.lua:44: in function 'forward'
	/home/amax/ceshi0415/daima/opensource_base.lua:528: in function 'train_epoch'
	opensource_baseline.lua:54: in function 'runTrainVal'
	opensource_baseline.lua:143: in main chunk
	[C]: at 0x004061f0

```
maybe error occurred when running the code---         local output = model:forward(input) in the following source code
   ```
 local filename = state.imglist[i]--'COCO_train2014_000000000092'
            local feat_visual = state.featureMap[filename]:clone()
            local feat_word = bagofword(manager_vocab, state.x_question[i])

            word_idx[nSample_batch] = state.x_question[i]
            featBatch_word[nSample_batch] = feat_word:clone()
            featBatch_visual[nSample_batch] = feat_visual:clone()
                
            while i == state.x_question:size(1) and nSample_batch< opt.batchsize do
                -- padding the extra sample to complete a batch for training
                nSample_batch = nSample_batch+1
                IDXset_batch[nSample_batch] = i
                target[nSample_batch] = first_answer
                featBatch_visual[nSample_batch] = feat_visual:clone()
                featBatch_word[nSample_batch] = feat_word:clone()
                word_idx[nSample_batch] = state.x_question[i]
            end 
            if nSample_batch == opt.batchsize then                
                nBatch = nBatch+1
                word_idx = word_idx:cuda()
                nSample_batch = 0
                target = target:cuda()
                featBatch_word = featBatch_word:cuda()
                featBatch_visual = featBatch_visual:cuda()
                ----------forward pass----------------------
                --switch between the baselines and the memn2n
                if opt.method == 'BOW' then
                    input = featBatch_word
                elseif opt.method == 'BOWIMG' then
                    input = {featBatch_word, featBatch_visual}
                elseif opt.method == 'IMG' then
                    input = featBatch_visual
                else 
                    print('error baseline method \n')
                end

                local output = model:forward(input)
                local err = criterion:forward(output, target)
                local prob_batch = output:float()
```

**Comment by richardassar** [comment]
The error is actually happening at `local err = criterion:forward(output, target)`, check your `target` tensor for values that are below 1 or greater than `output:size(2)` (assuming you are in batch mode).

**Comment by Larriers** [comment]
@richardassar 
Thank you. I'm new to the criterion and CNN. I want to know how can I suppose the `target`during test the dataset on a model that has finished the train.

**Comment by richardassar** [comment]
Sorry, you're best asking questions unrelated to this issue at https://groups.google.com/forum/#!forum/torch7

**Comment by purin0604** [comment]
I have the similar error with train.py from https://github.com/hughperkins/cifar.pytorch

> /tmp/luarocks_cunn-scm-1-9412/cunn/lib/THCUNN/ClassNLLCriterion.cu:57: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long *, Dtype *, int, int, int, int, long) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [19,0,0] Assertion `t >= 0 && t < n_classes` failed.
THCudaCheck FAIL file=/tmp/luarocks_cunn-scm-1-9412/cunn/lib/THCUNN/generic/ClassNLLCriterion.cu line=87 error=59 : device-side assert triggered
$ Error: cuda runtime error (59) : device-side assert triggered at /tmp/luarocks_cunn-scm-1-9412/cunn/lib/THCUNN/generic/ClassNLLCriterion.cu:87

Running with CUDA_LAUNCH_BLOCKING=1 returns the same error msg.

**Comment by brando90** [comment]
Wait what is going on? What does error mean?

**Comment by yxchng** [comment]
@richardassar so my target tensor cannot have 0? must start from 1?

**Comment by richardassar** [comment]
Lua uses 1-based indexing.

And also the documentation says:

"This criterion expects a class index (1 to the number of class) as target when
calling forward(input, target)
<https://github.com/torch/nn/blob/master/doc/criterion.md#nn.CriterionForward>
and backward(input, target)
<https://github.com/torch/nn/blob/master/doc/criterion.md#nn.CriterionBackward>
."

Please read the documentation.

**Comment by yxchng** [comment]
@richardassar ok thanks
